db/data.py

```diff
@@ -414,6 +454,83 @@ def load_high_level(mbid, offset=0):
         return {"metadata": metadata, "highlevel": highlevel}
 
 
+def load_many_high_level(recordings):
+    """Collect high-level data for multiple recordings.
+
+    Args:
+        recordings: A list of tuples (mbid, offset).
+
+    Returns:
+        {"mbid-1": {"offset-1": {"metadata-1": metadata, "highlevel-1": highlevel},
+                    ...
+                    ...
+                   "offset-n": {"metadata-n": metadata, "highlevel-n": highlevel}},
+
+        ...
+        ...
+
+        "mbid-n": {"offset-1": {"metadata-1": metadata, "highlevel-1": highlevel}}
+        }
+
+    """
+    with db.engine.connect() as connection:
+        # Metadata
+        meta_query = text("""
+            SELECT hl.id
+                 , hlm.data
+                 , ll.gid
+                 , ll.submission_offset
+              FROM highlevel hl
+         LEFT JOIN highlevel_meta hlm
+                ON hl.id = hlm.id
+              JOIN lowlevel ll
+                ON ll.id = hl.id
+             WHERE (ll.gid, ll.submission_offset)
+                IN :recordings
+        """)
+
+        meta_result = connection.execute(meta_query, { 'recordings': tuple(recordings) })
+        if not meta_result.rowcount:
+            raise db.exceptions.NoDataFoundException
+
+        hlids = []
+        recordings_info = defaultdict(lambda: defaultdict(lambda: defaultdict(dict)))
+        for row in meta_result.fetchall():
+            # Check that the hl calculation didn't fail (not just a placeholder row)
+            if row['data']:
+                hlids.append(row['id'])
+                recordings_info[str(row['gid'])][str(row['submission_offset'])]['metadata'] = row['data']
+
+        # Model data
+        model_query = text("""
+            SELECT m.model
+                 , hlmo.data
+                 , version.data as version
+                 , ll.gid
+                 , ll.submission_offset
+              FROM highlevel_model hlmo
+              JOIN model m
+                ON m.id = hlmo.model
+              JOIN version
+                ON version.id = hlmo.version
+              JOIN lowlevel ll
+                ON ll.id = hlmo.highlevel
+             WHERE hlmo.highlevel IN :hlids
+               AND m.status = 'show'
+        """)
+
+        model_result = connection.execute(model_query, { 'hlids': tuple(hlids) })
+        for row in model_result.fetchall():
+
+            model = row['model']
+            data = row['data']
+            data['version'] = row['version']
+
+            recordings_info[str(row['gid'])][str(row['submission_offset'])]['highlevel'][model] = data
+
+        return recordings_info
+
+
 def count_lowlevel(mbid):
```

Comment:
can you use the explain command in postgres (https://www.postgresql.org/docs/current/sql-explain.html) and paste the query plan for this query using both count() and max(submission_offset)? I'm not sure which one will be faster

Reply:
The query plan when using count():

<img width="588" alt="Screen Shot 2019-05-18 at 2 17 36 PM" src="https://user-images.githubusercontent.com/30450017/57973526-1a8d1200-7978-11e9-81bf-5652a07d62f1.png">

The query plan when using max(submission_offset) - we will also need to increment the offset:

<img width="600" alt="Screen Shot 2019-05-18 at 2 18 23 PM" src="https://user-images.githubusercontent.com/30450017/57973533-385a7700-7978-11e9-91ae-98f3f1d21754.png">

On my own, I'm not sure what to make of the difference between these.

Reply:
please always copy the output of programs as text, this way we can copy the contents if needed.

The cost for both of these queries is the same, so it doesn't look like there's a problem with what version we use.
I notice that your database is doing a sequential scan instead of an index scan, that's a bit strange as there should be an index in place. It's possible that you have so few rows that postgres decided not to use the index.

for example, mine shows:
```
                                     QUERY PLAN
-------------------------------------------------------------------------------------
 Aggregate  (cost=12.65..12.66 rows=1 width=8)
   ->  Bitmap Heap Scan on lowlevel  (cost=4.18..12.64 rows=4 width=0)
         Recheck Cond: (gid = 'bbe7bd1f-61c7-404d-811e-2bf42ca4b9cc'::uuid)
         ->  Bitmap Index Scan on gid_ndx_lowlevel  (cost=0.00..4.18 rows=4 width=0)
               Index Cond: (gid = 'bbe7bd1f-61c7-404d-811e-2bf42ca4b9cc'::uuid)
(5 rows)
```

Can you do the following things:
* copy the output of `\d lowlevel`, from a psql shell, and copy the output from the section that starts `Indexes:`, to verify that you have the necessary indexes.
* run `ANALYZE lowlevel;`, to recompute the query plans for this table.

Reply:
I believe my table didn't have the index because I was switching between branches and the other branches for #339 and #340 don't have the index migration yet. When I switch between working on each branch I typically reinitialize the database using `./develop.sh run --rm webserver python2 manage.py init-db --force` and import archived data. When I got these query plans in a psql session I must have been using the database as it was updated for one of the earlier steps.

A question I have about this... Is there a better way to update my database with the migrations that doesn't require me to reinitialize it and import my archived data? On the actual database for acousticbrainz.org, I have been wondering what the process is for making these changes as well. 

Anyways,  I reinitialized my database once on this branch and now my lowlevel table shows presence of the index,  and additionally my query plan matches yours:

```
                                          Table "public.lowlevel"
      Column       |           Type           | Collation | Nullable |               Default                
-------------------+--------------------------+-----------+----------+--------------------------------------
 id                | integer                  |           | not null | nextval('lowlevel_id_seq'::regclass)
 gid               | uuid                     |           | not null | 
 build_sha1        | text                     |           | not null | 
 lossless          | boolean                  |           |          | false
 submitted         | timestamp with time zone |           |          | now()
 gid_type          | gid_type                 |           | not null | 
 submission_offset | integer                  |           | not null | 
Indexes:
    "lowlevel_pkey" PRIMARY KEY, btree (id)
    "build_sha1_ndx_lowlevel" btree (build_sha1)
    "gid_ndx_lowlevel" btree (gid)
    "gid_type_ndx_lowlevel" btree (gid_type)
    "lossless_ndx_lowlevel" btree (lossless)
    "submission_offset_ndx_lowlevel" btree (submission_offset)
    "submitted_ndx_lowlevel" btree (submitted)
Referenced by:
    TABLE "highlevel" CONSTRAINT "highlevel_fk_lowlevel" FOREIGN KEY (id) REFERENCES lowlevel(id)
    TABLE "lowlevel_json" CONSTRAINT "lowlevel_json_fk_lowlevel" FOREIGN KEY (id) REFERENCES lowlevel(id)
```

```
                                     QUERY PLAN                                      
-------------------------------------------------------------------------------------
 Aggregate  (cost=12.65..12.66 rows=1 width=8)
   ->  Bitmap Heap Scan on lowlevel  (cost=4.18..12.64 rows=4 width=0)
         Recheck Cond: (gid = 'd1a0c57a-9b5c-4869-b039-629e845583fd'::uuid)
         ->  Bitmap Index Scan on gid_ndx_lowlevel  (cost=0.00..4.18 rows=4 width=0)
               Index Cond: (gid = 'd1a0c57a-9b5c-4869-b039-629e845583fd'::uuid)
```